core/src/main/scala/org/apache/spark/TestUtils.scala

```diff
@@ -448,6 +451,33 @@ private[spark] object TestUtils {
       EnumSet.of(OWNER_READ, OWNER_EXECUTE, OWNER_WRITE))
     file.getPath
   }
+
+  /**
+   * A log4j-specific log capturing mechanism. Provided with a class name, it will add a new
+   * appender to the log for that class to capture the output. log4j must be properly configured
+   * on the classpath (e.g., with slf4j-log4j12) for this to work. This should be closed when it is
+   * done to remove the temporary appender.
+   */
+  class Log4jCapture(val loggerName: String) extends AutoCloseable {
+    private val logger = LogManager.getLogger(loggerName)
+    private val writer = new StringWriter()
+    private val appender = new WriterAppender(new PatternLayout(), writer)
+    logger.addAppender(appender)
+
+    def this(classToCapture: Class[_]) = {
+      this(classToCapture.getName)
+    }
+
```

Comment:
seems this constructor isn't used? Maybe we can move the logic to strip the `$`-suffix here

Reply:
Done